libvcx/src/utils/devsetup_agent.rs

```diff
@@ -0,0 +1,462 @@
+    use crate::libindy::utils::wallet::*;
+    use crate::utils::devsetup::*;
+    use crate::utils::provision::{provision_cloud_agent, ProvisionAgentConfig, AgencyConfig};
+    use crate::init::{open_as_main_wallet, init_issuer_config, create_agency_client_for_main_wallet};
+    use crate::utils::constants;
+
+    #[derive(Debug)]
+    pub struct VcxAgencyMessage {
+        pub uid: String,
+        pub decrypted_msg: String,
+    }
+
+    fn determine_message_type(a2a_message: A2AMessage) -> PayloadKinds {
+        debug!("determine_message_type >>> a2a_message={:?}", a2a_message);
+        match a2a_message.clone() {
+            A2AMessage::PresentationRequest(_) => PayloadKinds::ProofRequest,
+            A2AMessage::CredentialOffer(_) => PayloadKinds::CredOffer,
+            A2AMessage::Credential(_) => PayloadKinds::Cred,
+            A2AMessage::Presentation(_) => PayloadKinds::Proof,
+            _msg => PayloadKinds::Other(String::from("aries"))
+        }
+    }
+
+    fn str_message_to_a2a_message(message: &str) -> VcxResult<A2AMessage> {
+        Ok(serde_json::from_str(message)
+            .map_err(|err| VcxError::from_msg(VcxErrorKind::InvalidJson, format!("Cannot deserialize A2A message: {}", err)))?
+        )
+    }
+
+    fn str_message_to_payload_type(message: &str) -> VcxResult<PayloadKinds> {
+        let a2a_message = str_message_to_a2a_message(message)?;
+        Ok(determine_message_type(a2a_message))
+    }
+
+    fn download_message(did: String, filter_msg_type: PayloadKinds) -> VcxAgencyMessage {
+        let mut messages = agency_client::get_message::download_messages_noauth(Some(vec![did]), Some(vec![String::from("MS-103")]), None).unwrap();
+        assert_eq!(1, messages.len());
+        let messages = messages.pop().unwrap();
+
+        for message in messages.msgs.into_iter() {
+            let decrypted_msg = &message.decrypted_msg.unwrap();
+            let msg_type = str_message_to_payload_type(decrypted_msg).unwrap();
+            if filter_msg_type == msg_type {
+                return VcxAgencyMessage {
+                    uid: message.uid,
+                    decrypted_msg: decrypted_msg.clone(),
+                };
+            }
+        }
+        panic!("Message not found")
+    }
+
+    pub trait TestAgent {
+        fn activate(&mut self) -> VcxResult<()>;
+    }
+
+    pub struct Faber {
+        pub is_active: bool,
+        pub config_wallet: WalletConfig,
+        pub config_agency: AgencyConfig,
+        pub config_issuer: IssuerConfig,
+        pub connection_handle: u32,
+        pub schema_handle: u32,
+        pub cred_def_handle: u32,
+        pub credential_handle: u32,
+        pub presentation_handle: u32,
+    }
+
+    impl TestAgent for Faber {
+        // propagte error up so we can handle it gracefully in
+        /// impl Drop for Faber
+        // so that we don't double panic, in case we are dropping due to a panick in the test
```

Comment:
1.) "propagate"
2.) //
3.) "panic"
4.) I am sorry
5.) I guess this should be a todo?

Reply:
yes this was actual todo during while working on this, I must had been in real hurry when I was typing this :-D Anyway the stuff which is contained there is implemented.
There was a problem that when test panicked, consequently `Drop` also panicked, causing double panic, making it quite hard to troubleshoot. It doesn't happen anymore, because `Drop` now never panics, but rather handle any error which might occur during the `drop`
```
fn drop(&mut self) {
            self.activate().unwrap_or_else(|e| error!("Failed to close main wallet while dropping Faber"));
            close_main_wallet().unwrap_or_else(|e| error!("Failed to close main wallet while dropping Faber"));
            delete_wallet(&self.config_wallet).unwrap_or_else(|e| error!("Failed to delete Faber's wallet while dropping"));
        }
 ```

Reply:
I managed to understand that from the comment, but didn't find drop